### ISSUE TITLE
do not open all files at once when doing add (Too many open files)

### ISSUE THREAD

**Issue by alefminus** [Oxen-AI/oxen-archive#93]
Reproduce:
```
mkdir too-many-files
cd too-many-files
ulimit -n 50
for ((i=0;i<52;++i)); do touch test$i.txt; done
oxen init
oxen add *.txt
```
Output
```
██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 0/0
Commit 9f540298cd2ae0ff done.
Initial commit 9f540298cd2ae0ff
🐂 repository initialized at: "/home/alon/src/Oxen/bugs/too-many-files"
DB(Error { message: "IO error: While opening a file for sequentially reading: /home/alon/src/Oxen/bugs/too-many-files/.oxen/history/9f540298cd2ae0ff/files//CURRENT: Too many open files" })

```

**Comment by luanmuniz** [comment]
I had a similar issue. For me was when i did `oxen add .` Here is the log:

```
Modified files:    (use "oxen add <file>..." to update what will be committed)
  modified: download-full-repo.js

Untracked Directories
  (use "oxen add <dir>..." to update what will be committed)
  data/ (22 items)

cvm-data-repo: $ oxen add .
Adding files in directory: "data"
██░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 20/2111thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DB(Error { message: "IO error: While open a file for appending: ~/cvm-data-repo/.oxen/staged/files/data/FIE/DOC/BALANCETE/DADOS/balancete_fie_202001/000009.dbtmp: Too many open files" })', src/lib/src/index/stager.rs:585:77
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DB(Error { message: "IO error: While open a file for appending: ~/cvm-data-repo/.oxen/history/43e6309b1265ba99/files/data/FIE/DOC/BALANCETE/DADOS/HIST/balancete_fie_201406/LOG: Too many open files" })', src/lib/src/index/stager.rs:588:22
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DB(Error { message: "IO error: While open a file for appending: ~/cvm-data-repo/.oxen/history/43e6309b1265ba99/files/data/FIE/DOC/BALANCO/DADOS/balanco_fie_2013/MANIFEST-000005: Too many open files" })', src/lib/src/index/stager.rs:588:22
[2023-03-14T14:58:14Z ERROR liboxen::index::stager] Could not add file: "data/CIA_ABERTA/DOC/ITR/META/meta_itr_cia_aberta_txt/meta_itr_cia_aberta_DFC_MI.txt"
    Err: util::hasher::hash_file_contents Could not open file "~/cvm-data-repo/data/CIA_ABERTA/DOC/ITR/META/meta_itr_cia_aberta_txt/meta_itr_cia_aberta_DFC_MI.txt"
██░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 22/2111thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DB(Error { message: "IO error: While opendir: ~/cvm-data-repo/.oxen/staged/files/data/FIE/DOC/BALANCETE/DADOS/HIST/balancete_fie_201308: Too many open files" })', src/lib/src/index/stager.rs:585:77
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DB(Error { message: "Invalid argument: Found options incompatible with filesystem: IO error: While open a file for random read: ~/cvm-data-repo/.oxen/staged/files/data/FIE/DOC/BALANCETE/DADOS/balancete_fie_201804/CURRENT: Too many open files" })', src/lib/src/index/stager.rs:585:77
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DB(Error { message: "IO error: While open a file for appending: ~/cvm-data-repo/.oxen/staged/files/data/FIE/DOC/BALANCETE/DADOS/HIST/balancete_fie_201203/000009.dbtmp: Too many open files" })', src/lib/src/index/stager.rs:585:77
fatal runtime error: Rust cannot catch foreign exceptions
Abort trap: 6
````

I end up managing to push all the files by adding, committing and pushing little by little. This is the final repo https://www.oxen.ai/luanmuniz/cvm

**Comment by gschoeni** [comment]
Looking into it 👍

**Comment by gschoeni** [comment]
Added an environment variable for MAX_OPEN_FILES, and a default of 128 to the code to fix this. Tested with the commands above.

**Comment by alefminus** [comment]
excellent, thanks for taking care of this!

On Saturday, April 8, 2023, Greg Schoeninger ***@***.***>
wrote:
> Closed #93 as completed.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.<
https://ci4.googleusercontent.com/proxy/DfPmtXsV6D_ak-gmoIlRqwqw_bxWKjxT20iCTGKY0CF7oXrPN1rpLuBR-t5BcsUq7i602pO98BXvlpEshFnr2f_E7Q_FpKE9fNUGutcna4V0QAvuv2OJFDn2AqCt79M_ezrfUGHmm8T_DN5P7X29LuoVREaYYaFdTiaFZayEafzsCZWEIoYk7YUWDaR0mAT1-JI7cOlskt3NgBl9KWLebRc4B3b214y2huq-7KJqhaesdoYcOIUgopm6jKEuZjshXeEO0hw-_4turstO=s0-d-e1-ft#https://github.com/notifications/beacon/AYWFZPDXN5TVHDBXA7MESALXAC34BA5CNFSM6AAAAAAVCWEQXCWGG33NNVSW45C7OR4XAZNWJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XKUY3PNVWWK3TUL5UWJTYAAAAAEFO52QTQ.gif>Message
ID: ***@***.***>

-- 
Alon Levy
Software Developer
Greenvibe
054-2395317
